lib/rtti.h

```diff
@@ -0,0 +1,181 @@
+#elif defined(__GNUC__)
+    return __PRETTY_FUNCTION__;
+#else
+#error "Unsupported compiler"
+#endif
+}
+
+static constexpr uint64_t FNV1a(const char *str, size_t n,
+                                uint64_t hash = UINT64_C(0xcbf29ce484222325)) {
+    return n == 0 ? hash : FNV1a(str + 1, n - 1, UINT64_C(0x100000001b3) * (hash ^ str[0]));
+}
+
+static constexpr uint64_t FNV1a(TypeNameHolder str) { return FNV1a(str.str, str.length); }
+
+// Default implementation of typeid resolver: use hash derived from type name,
+// but mark upper 8 bits to be 0xFF
+template <typename T, typename = void>
+struct TypeIdResolver {
+    static constexpr TypeId resolve() noexcept {
+        // Get original 64-bit hash
+        uint64_t hash = FNV1a(TypeNameProxy<T>());
+        // Xor-fold to obtain 56-bit value
+        hash = (hash >> 56) ^ (hash & ((UINT64_C(1) << 56) - 1));
+        return (UINT64_C(0xFF) << 56) | hash;
+    }
+};
+
+// Specialized implementation that looks for static_typeId member.
+template <typename T>
+struct TypeIdResolver<T, std::void_t<decltype(T::static_typeId)>> {
+    static constexpr uint64_t resolve() noexcept {
+        constexpr TypeId id = T::static_typeId();
+        // Allow fall-back implementation if class does not want to explicitly
+        // specify a typeid
+        if constexpr (id == InvalidTypeId) return TypeIdResolver<T, int>::resolve();
+
+        return id;
+    }
+};
+
+}  // namespace detail
+
+struct Base;
+
+template <typename This, typename... Parents>
+struct TypeInfo {
+    using T = std::remove_const_t<std::remove_reference_t<This>>;
+
+    static_assert((... && std::is_base_of_v<Parents, This>),
+                  "One or more parents are not a base of this type.");
+
+    static_assert((... && std::is_base_of_v<Base, Parents>),
+                  "One or more parent hierarchies is not based on top of RTTI::Base.");
+
+    static_assert(std::is_base_of_v<Base, This>,
+                  "Invalid typeinfo requested, class is not based on top of RTTI::Base.");
+
+    static_assert(
+        std::is_same_v<T *, decltype(T::rttiEnabledMarker(nullptr))>,
+        "Invalid typeinfo requested, class does not declare typeinfo via DECLARE_TYPEINFO.");
+
+    [[nodiscard]] static constexpr TypeId id() noexcept {
+        return detail::TypeIdResolver<T>::resolve();
+    }
+
+    [[nodiscard]] static constexpr bool isA(TypeId typeId) noexcept {
+        return (id() == typeId) || (... || (Parents::TypeInfo::isA(typeId)));
+    }
+
+    template <typename T>
+    [[nodiscard]] static const void *dyn_cast(TypeId typeId, const T *ptr) noexcept {
+        if (id() == typeId) return static_cast<const This *>(ptr);
+
+        std::array<const void *, sizeof...(Parents)> ptrs = {
+            Parents::TypeInfo::dyn_cast(typeId, ptr)...};
+
+        auto it =
+            std::find_if(ptrs.begin(), ptrs.end(), [](const void *ptr) { return ptr != nullptr; });
+        return (it != ptrs.end()) ? *it : nullptr;
```

Comment:
Why would this not work:

```cpp
if (isA(typeId)) return static_cast<const This *>(ptr);
```

I don't see why you would need to defer to the parent's cast function. Also I think the `Parents::TypeInfo` would cast it to the parent pointer, which might need adjusting in presence of multiple inheritance.

Also, this is definitely one of the places that could use comments what exactly is the meaning of the arguments and the return type.

Reply:
See below